### 一. 基础知识练习
___

#### 1. 函数的定义和调用

In [3]:
# 定义一个名称为 draw_rect 的函数，功能是绘制指定大小（高度和宽度）的星号矩形，高度和宽度是绘制矩形的参数
def draw_react(height,width):
    for i in range(height):
        print("*" * width)

# 调用 draw_rect 函数，绘制高度为 5，宽度为 10 的星号矩形
draw_react(5,10) 

**********
**********
**********
**********
**********


#### 2. 参数的传递

##### 2.1 位置参数、关键字参数

In [10]:
def draw_rect(width,height,star):
    for i in range(height):
        print(star*width)

# 位置参数
draw_rect(5,3,"*")
print()

# 关键字参数
draw_rect(width=5,height=3,star="*")
print()

# 位置参数和关键字参数混合使用
draw_rect(5,height=3,star="*")

# 错误示范
# draw_rect(width=5,3,"*")

*****
*****
*****

*****
*****
*****

*****
*****
*****


#### 2.2 可选参数（默认值参数）

In [12]:
def draw_react3(width,height,star = '*'):
    for i in range(height):
        print(star * width)

# 不写第三个实参
draw_react3(5,6)
print()
# 写第三个实参
draw_react3(5,6,'#')

*****
*****
*****
*****
*****
*****

#####
#####
#####
#####
#####
#####


##### 2.3 变长参数

In [16]:
def order_dishes(customer,*dishes,num_people,**addtional_comments):
    print(f"欢迎您！{customer}")
    print(f"您预定的菜单为：{dishes}")
    print(f"进餐人数是{num_people}人")
    print(f"您附加的要求为{addtional_comments}")

# 模拟订餐
order_dishes("张先生","土豆炖牛腩","奥尔良烤翅","家常豆腐",num_people=3,口味="偏清淡",餐桌位置="偏僻角落")

# 说明：关键字参数的键必须是变量名，而不是字符串，但是输出后会自动加上引号

欢迎您！张先生
您预定的菜单为：('土豆炖牛腩', '奥尔良烤翅', '家常豆腐')
进餐人数是3人
您附加的要求为{'口味': '偏清淡', '餐桌位置': '偏僻角落'}


##### 2.4 解包参数


In [18]:
# 利用 * 将列表或元组解包为多个参数
def func1(a,b,c):
    print(a,b,c)
    return a + b + c

args = [1,2,3]
res = func1(*args)
print(res)

1 2 3
6


In [19]:
# 利用 ** 将字典键值对解包为多个关键字参数
def func2(x,y,z):
    print(x,y,z)
    return x + y + z

kargs = {"x":1,"y":2,"z":3}
res = func2(**kargs)
print(res)

1 2 3
6


##### 2.6 传递列表

a. 一个对照试验


In [32]:
# 定义函数increment_count
def increment_count(count:int, increment:int):
    """使count的值增加increment"""
    print(f"执行增量前，count的值是{count}, 标识是{id(count)}")
    count += increment
    print(f"执行增量后，count的值是{count}, 标识是{id(count)}")

# 定义函数increment_counts
def increment_counts(counts:list, increment:int):
    """使counts中所有元素的值增加increment"""
    print(f"执行增量前，counts的值是{counts}, 标识是{id(counts)}")
    for i in range(len(counts)):
        counts[i] += increment
    print(f"执行增量后，counts的值是{counts}, 标识是{id(counts)}")

# 对比实验
count_value = 100
print(f"调用增量函数前，count_value的值是{count_value}, 标识是{id(count_value)}")
increment_count(count_value, 10) # 把整数100的引用传递给形参count
print(f"调用增量函数后，count_value的值是{count_value}, 标识是{id(count_value)}")
print()

count_values = [100, 200, 300]
print(f"调用增量函数前，count_values的值是{count_values}, 标识是{id(count_values)}")
increment_counts(count_values, 10) # 把列表对象[100, 200, 300]的引用传递给形参counts
print(f"调用增量函数后，count_values的值是{count_values}, 标识是{id(count_values)}")


调用增量函数前，count_value的值是100, 标识是140729707804184
执行增量前，count的值是100, 标识是140729707804184
执行增量后，count的值是110, 标识是140729707804504
调用增量函数后，count_value的值是100, 标识是140729707804184

调用增量函数前，count_values的值是[100, 200, 300], 标识是2318735827904
执行增量前，counts的值是[100, 200, 300], 标识是2318735827904
执行增量后，counts的值是[110, 210, 310], 标识是2318735827904
调用增量函数后，count_values的值是[110, 210, 310], 标识是2318735827904


说明：
- 整数 100 是不可变对象，函数操作对100没有影响，也不会改变 `count_value` 对100的引用，执行增量后，`count` 引用了一个新对象，其值为110，其标识不同于100的标识
- 列表 [100,200,300] 为可变对象，函数操作对其有影响，执行增量前后，`count_values` 和 `counts` 引用的都是这个列表对象

b. 随机混排给定列表的元素值

In [33]:
import random
def shuffle(lst):
    for i in range(len(lst)):
        j = random.randint(0,len(lst)-1)
        lst[i],lst[j] = lst[j],lst[i]

# 对shuffle函数进行测试
def test_shuffle():
    v = [1,2,3,4,5]
    shuffle(v)
    print(v)

test_shuffle()

[5, 1, 4, 3, 2]


c. 来访者登记簿程序：定义函数 `check_in` 登记来访者的姓名和电话，并把记录保存到一个来访者列表


In [34]:
def check_in(visitors:list):
    """记录来访者输入的姓名和联系电话，追加到visitors"""
    name = input("请输入姓名：")
    tel = input("联系电话：")
    visitors.append((name,tel))

visitor_book = []    # 定义列表用来保存即将到来的来访者信息

# 假设有 3 名来访者的信息要登记，'_'用作临时循环变量
for _ in range(3):
    check_in(visitor_book)

print(visitor_book)

[('沙', '123456'), ('夏', '234567'), ('灿', '345678')]


#### 3. 函数的返回值

##### 3.1 返回简单值
定义函数circle_area，根据半径计算圆的面积

In [22]:
# 定义函数circle_area，根据半径计算圆的面积
import math

def circle_area(r):
    return math.pi*r**2

radius = eval(input("请输入圆的半径为：r = "))
area = circle_area(radius)
print(f"半径为 {radius} 的圆的面积为 {area:.2f}")


半径为 4 的圆的面积为 50.27


##### 3.2 返回逻辑值
定义函数 is_even 检测任意一个大于零的整数是否为整数，返回检测结果，是一个布尔型的逻辑值

In [24]:
# 定义函数 is_even 检测任意一个大于零的整数是否为整数，返回检测结果，是一个布尔型的逻辑值
def is_even(n):
    if n > 0 and n % 2 == 0:
        return True
    else:
        return False

num = int(input("请输入一个大于零的偶数："))
if is_even(num):
    print(f"{num} 是大于 0 的偶数")
else:
    print("不满足条件：不是大于零的偶数")

4 是大于 0 的偶数


##### 3.3 包含多个 `return` 语句的函数
定义函数is_prime判断一个数是否为素数，编写测试代码判断并输出1~99中的素数

In [28]:
# 定义函数is_prime判断一个数是否为素数，编写测试代码判断并输出1~99中的素数
def is_prime(n):
    if n < 2:
        return False
    elif n == 2:
        return True
    else:
        i = 2
        while i <= n // 2 + 1:
            if n % i == 0:
                return False
            else:
                i += 1
        return True

prime_lst = []
for j in range(1,100):
    if is_prime(j):
        prime_lst.append(j)

print(f"1~99 内的素数有 {','.join(map(str,prime_lst))}")


1~99 内的素数有 2,3,5,7,11,13,17,19,23,29,31,37,41,43,47,53,59,61,67,71,73,79,83,89,97


##### 3.4 返回元组（包含多个值）
定义一个函数describe，计算一组数的 3 个基本描述性统计量（最大值、最小值和平均值），返回这三个统计量

In [31]:
# 定义一个函数describe，计算一组数的 3 个基本描述性统计量（最大值、最小值和平均值），返回这三个统计量
def describe(*args):
    maximum = max(args)
    minimum = min(args)
    medium = sum(args)/len(args)
    return maximum,minimum,medium

lst = eval(input("请输入一组数据："))
print(lst)

describe_tup = describe(*lst)

print(f"最大值为 {describe_tup[0]}")
print(f"最小值为 {describe_tup[1]}")
print(f"平均值为 {describe_tup[2]}")

(1, 2, 3, 4)
最大值为 4
最小值为 1
平均值为 2.5


#### 4. 函数式编程（综合利用纯函数、高阶函数）
##### 计算器模拟程序

In [1]:
# 定义加减乘除的纯函数
def add(x,y):
    """加法运算 + """
    return x + y
def substract(x,y):
    """减法运算"""
    return x - y
def multiply(x,y):
    """乘法运算"""
    return x * y
def divide(x,y):
    """乘法运算"""
    return x/y


# 定义纯函数和高阶函数

# 返回运算符
def calc_function(operator_code):
    """根据operator_code的值返回相应运算的函数"""
    functions = [add,substract,multiply,divide]
    return functions["+-*/".index(operator_code)]

def calc(function,x,y):
    """function接收函数参数"""
    return function(x,y)

def calculator(operator_code,x,y):
    """根据operator_code的值完成x和y的运算"""
    function = calc_function(operator_code)
    result = calc(function,x,y)
    return result


# 主程序——计算12和34的和

# 1.利用纯函数直接计算
print(add(12,34))

# 2. 将add函数赋值给变量add_function中
add_function = add
print(add_function(12,34))    # 该变量可以作为函数名来调用

# 3. 利用calc_function
print(calc_function("+")(12,34))

# 4. 利用calc,add做实参
print(calc(add,12,34))
  
# 5. 利用calculator
print(calculator("+",12,34))


46
46
46
46
46


#### 5.利用内置函数进行函数式编程
##### 5.1 利用 `map()` 函数
a. 输入一个三位数，把每一位上的数分别赋值给a,b,c


In [5]:
x = input("请输入一个三位数：")
a,b,c = map(int,x)
print(a,b,c)

# map对象不能通过print()函数显示元素值，可以调用list()函数把它转换为列表对象后输出
print(map(list,x))
lst = list(map(int,x))
print(lst)

7 8 9
[7, 8, 9]


b. 利用 `map()` 函数计算一个数据集中每项数据的平方

In [6]:
def square(x):
    return x*x

lst = [1,2,3,4,5,6]

square_lst = list(map(square,lst))
print(square_lst)

[1, 4, 9, 16, 25, 36]


c. 多个可迭代对象

In [8]:
def add(x,y):
    return x + y

lst1 = [1,2,3,4]
lst2 = [6,7,8,9]

lst3 = list(map(add,lst1,lst2))
print(lst3)

[7, 9, 11, 13]


#### 5.2 `lambda`表达式（匿名函数）
a. 按照学生总成绩对列表进行排序

In [12]:
# 保存5位学生的姓名和3门课成绩
score_sheet = [
    ['李华', 85, 87, 80],
    ['陈琪', 93, 91, 88],
    ['余涛', 88, 90, 87],
    ['郑燕', 82, 78, 80],
    ['金浩', 91, 88, 85]]

# 按学生的成绩总分从高到低的顺序排序
ordered_sheet = sorted(score_sheet,reverse = True,key = lambda x:sum(x[1:]))
print(ordered_sheet)

# 利用max函数从成绩表中选出总分最高的学生
highest_student = max(score_sheet,key = lambda x:sum(x[1:]))
print(highest_student)

[['陈琪', 93, 91, 88], ['余涛', 88, 90, 87], ['金浩', 91, 88, 85], ['李华', 85, 87, 80], ['郑燕', 82, 78, 80]]
['陈琪', 93, 91, 88]


b. 生成一个包含多个列表(元素为从1到10的整数)的列表，并按照子列表的第二个和第五个数字的升序对该列表进行排序

In [ ]:
import random

lst = [[random.randint(1,10) for i in range(5)] for j in range(5)]

for row in lst:
    print(row)
print()

ordered_lst = sorted(lst,reverse = False,key = lambda x:(x[1],x[4]))
for line in ordered_lst:
    print(line)

# 注意：按两个元素进行升序，首先比较x[1],若x[1]相同，再比较x[4]

[5, 1, 6, 6, 1]
[7, 4, 7, 4, 5]
[8, 9, 8, 6, 6]
[6, 5, 4, 2, 5]
[9, 10, 2, 5, 9]

[5, 1, 6, 6, 1]
[7, 4, 7, 4, 5]
[6, 5, 4, 2, 5]
[8, 9, 8, 6, 6]
[9, 10, 2, 5, 9]


##### 5.3 `filter()` 函数
用 `filter()` 函数找出平均分大于或等于85分的学生


In [17]:
# 保存5位学生的姓名和3门课成绩
score_sheet = [
    ['李华', 85, 87, 80],
    ['陈琪', 93, 91, 88],
    ['余涛', 88, 90, 87],
    ['郑燕', 82, 78, 80],
    ['金浩', 91, 88, 85]]

student_filter = filter(lambda x:sum(x[1:])/(len(x)-1)>85,score_sheet)
print(list(student_filter))

[['陈琪', 93, 91, 88], ['余涛', 88, 90, 87], ['金浩', 91, 88, 85]]


##### 5.4 `reduce()` 函数
利用函数 `reduce()` 计算一组数的乘积


In [18]:
from functools import reduce 

lst = [1,2,3,4,5]

m = reduce(lambda x,y:x*y,lst)
print(m)

120


#### 6. 递归函数
##### 6.1 定义一个简单的递归函数，计算斐波那契数列的第n项，其中第0项是0，第一项是1，第二项后每一项为其前一项之和

In [21]:
def fib(n):
    """用递归函数的方法计算斐波那契数列的第n项"""
    if n == 0 or n == 1:      # 递归基（基本情形）
        return  n
    else:
        return fib(n-1) + fib(n-2)

# 计算第十项
fib_10 = fib(10)
print(fib_10)
# 计算前10项之和
fib_10_sum = sum([fib(i) for i in range(1,11)])
print(fib_10_sum)

55
143


##### 6.2 汉诺塔问题
递归步骤的核心思路\
当 n > 1 时，要把 n 个圆盘从 x 移动到 z，可以分解为三个子任务：
- 先移动上面的 n-1 个圆盘：把它们从起始柱 x 移动到辅助柱 y（借助目标柱 z 作为临时辅助）。
- 移动最大的圆盘：把第 n 个圆盘（最大的那个）从起始柱 x 直接移动到目标柱 z。
- 再移动剩下的 n-1 个圆盘：把之前暂存在辅助柱 y 上的 n-1 个圆盘，移动到目标柱 z（借助起始柱 x 作为临时辅助）。

In [26]:
# 定义圆盘移动函数
def move(di,s,d):
    """把编号为di的圆盘从塔s移动到d"""
    print(f"{di}\t{s}-->{d}")

# 定义求解汉诺塔问题的递归函数
def hanoi(n,x,y,z):
    if n == 1:
        move(n,x,z)
    else:
        hanoi(n-1,x,z,y)
        move(n,x,z)
        hanoi(n-1,y,x,z)

hanoi(3,"X","Y","Z")

1	X-->Z
2	X-->Y
1	Z-->Y
3	X-->Z
1	Y-->X
2	Y-->Z
1	X-->Z


##### 6.3 用递归的方式来反转字符串

In [27]:
def reverse_string(s):
    if len(s) <= 1:
        return s
    return reverse_string(s[1:]) + s[0]

print("Hello")
print(reverse_string("Hello"))

Hello
olleH


##### 6.4 用递归法则对整数进行质因数分解

In [34]:
import random

def factors(num,fac = []):
    for i in range(2,int(num/2)+1):
        if num % i == 0:
            fac.append(i)
            factors(num//i,fac)
            break
    else:
        fac.append(num)
factors_lst = []
n = random.randint(2,100)
print(n)
factors(n,factors_lst)
print(factors_lst)

21
[3, 7]


### 二. 实践作业
___

#### 1. 编写程序，定义一个求斐波那契数列的函数fib(n)，并编写测试代码，输出前20项。要求一行输出10项，每项至少占5个字符位置，右对齐。

In [38]:
def fib(n):
    """用于生成斐波那契数列,参数n表示数列的项数"""
    H_i = []
    for i in range(1,n + 1):
        if i == 1 or i == 2:
            H_i.append(1)
        else:
            H_i.append(H_i[i-2]+H_i[i-3])
    return H_i

num_seq = int(input("请输入要输出的斐波那契数列的项数：n = "))
fib_lst = fib(num_seq)
for j in fib_lst:
    print(f"{j:>5}",end = '')
    if (fib_lst.index(j)+1) % 10 == 0:
        print()



    1    1    2    3    5    8   13   21   34   55
   89  144  233  377  610  987 1597 2584 4181 6765


In [42]:
def fib2(n):
    """用于生成斐波那契数列的第n项"""
    if n == 0 or n == 1:
        return n
    else:
        return fib2(n-1) + fib2(n-2)

num_seq = int(input("请输入要输出的斐波那契数列的项数：n = "))
fib_lst = [fib2(i) for i in range(1,num_seq + 1)]
count = 0
for j in fib_lst:
    count += 1
    print(f"{j:>5}",end = '')
    if count % 10 == 0:
        print()

    1    1    2    3    5    8   13   21   34   55
   89  144  233  377  610  987 1597 2584 4181 6765


#### 2. 编写程序，定义一个函数IsPerfectNumber(n)判断一个整数是否是完全数。如果一个整数的所有因子的和等于该整数，它就是完全数。用该函数找出1~1000内的所有完全数。

In [44]:
def IsPerfectNumber(n):
    """用于判断一个数是否是完全数"""
    factors = [] 
    for i in range(1,int(n/2) + 1):
        if n % i == 0:
            factors.append(i)
    if sum(factors) == n:
        return True
    else:
        return False

perfect_lst = [i for i in range(1,1001) if IsPerfectNumber(i)]
print(perfect_lst)

[6, 28, 496]


#### 3. 编写程序，利用可变参数定义一个求任意2个或更多数中最小值的函数min_n(a, b, *c)，并编写测试代码。
例如，print(min_n(8, 2))，print(min_n(16, 1, 7, 4, 15))

In [45]:
def min_n(a,b,*c):
    """求出最小值"""
    if a > b:
        a = b
    for i in c:
        if a > i:
            a = i
    return a

print(min_n(8,2))
print(min_n(16,1,7,4,15))

2
1


#### 4. 编写程序，利用元组作为函数的返回值，求序列数据中的最大值、最小值和元素个数，并编写测试代码。假设测试数据分别为s1=[9,7,8,3,2,1,55,6]，s2=["apple", "pear", "meton", "kiwi"]和s3="TheQuickBrownFox"。运行效果如下：
  

list=[9,7,8,3,2,1,55,6]\
最大值=55, 最小值=1, 元素个数=8\
list=["apple", "pear", "meton", "kiwi"]\
最大值=pear, 最小值=apple, 元素个数=4\
list=TheQuickBrownFox\
最大值=x, 最小值=B, 元素个数=16

In [46]:
def describe(lst):
    """序列数据中的最大值、最小值和元素个数"""
    max_value = max(lst)
    min_value = min(lst)
    num = len(lst)
    return max_value,min_value,num

# 定义测试函数
def test():
    # 定义三个测试序列
    s1 = [9,7,8,3,2,1,55,6]
    s2 = ["apple", "pear", "meton", "kiwi"]
    s3 = "TheQuickBrownFox"
    
    # 将三个测试序列存储在一个列表s中
    s = [s1,s2,s3]
    
    # 遍历列表s中的每个序列
    for i in s:
        # 调用basic_infor函数，获取当前序列的最大值、最小值和元素个数
        a,b,c = describe(i)
        # 打印当前序列
        print(f"list = {i}")
        # 打印序列的最大值、最小值和元素个数
        print("最大值 = {0}，最小值 = {1}，元素个数 = {2}".format(a,b,c))

# 调用测试函数test
test()

list = [9, 7, 8, 3, 2, 1, 55, 6]
最大值 = 55，最小值 = 1，元素个数 = 8
list = ['apple', 'pear', 'meton', 'kiwi']
最大值 = pear，最小值 = apple，元素个数 = 4
list = TheQuickBrownFox
最大值 = x，最小值 = B，元素个数 = 16


#### 5. 编写程序，定义一个查找函数search(dataset, k)，在 dataset 列表中查找指定的数 k 出现的所有位置，然后编写测试代码，产生一组随机整数，由用户输入要找的数，使用search函数查找，并输出查找结果。

In [49]:
import random

def search(dataset,k):
    """
    在 dataset 列表中查找指定数字 k 出现的所有位置。

    :param dataset: 列表，包含随机整数
    :param k: 要查找的数字
    :return: None,直接打印结果
    """
    location = []
    for i in range(len(dataset)):
        if dataset[i] == k:
            location.append(i + 1)
    if location:
        print(f"{k}出现在第{",".join(map(str,location))}个位置")
    else:
        print(f"{k}不在该列表之中")


lower,upper = eval(input("请输入随机数字列表的范围（最小值和最大值，用“,”隔开）:"))
length = int(input("请输入列表的长度为："))
n = int(input("请输入你要查找的数字："))
lst = [random.randint(lower,upper) for i in range(length)]
print(lst)
search(lst,n)


[37, 22, 89, 66, 20, 94, 86, 47, 24, 80, 80, 22, 97, 50, 47, 85, 48, 40, 32, 63]
88不在该列表之中


#### 6. 编写并测试一个递归函数mymax，它找出列表中最大的数字。
  提示：mymax是第一个数据项和所有其他数据项的最大值中较大的一个。例如，mymax([23,12,64,8,39])的返回值为64。

In [50]:
def mymax(lst:list):
    """
    使用递归方法查找列表 lst 中的最大值。

    :param lst: 列表，包含数字
    :return: 列表中的最大值
    """
    if len(lst) == 1:
        return lst[0]
    else:
        other_max = mymax(lst[1:])
        return lst[0] if lst[0] > other_max else other_max
    
test_lst = [23,12,64,8,39]
print(mymax(test_lst))
    

64


#### 7. “回文”是顺着读或倒着读含有相同顺序的字母的句子（不包含标点符号），一个典型的例子是“Able was I, ere I saw Elba”。写一个函数来检测一个字符串是不是回文。
  提示：如果打算用递归方法，其基本思想是检查字符串的第一个和最后一个字母是否相同；如果相同，那么如果这两个字母之间的所有内容都是回文，它就是回文。有两种特殊情况需要处理：\
（1）如果字符串中包含有标点符号，需要检查该字符串删除标点；\
（2）在比较字母时，请确保不区分大小写。在程序中使用你的函数，提示用户输入短语，然后指出它是不是回文。另一个经典的测试是“A man, a plan, a canal, Panama!”

#### 8. 计算机科学家和数学家经常使用 10 以外基数的进制系统。编写一个程序，允许用户输入一个数字和一个基数，然后打印出新基数中的数字。使用递归函数 baseConversion(num, base)打印数字。
  提示：以基数为 10（十进制）的数字为例，要获得基数 为10 的数字的最右边的数字，只需除以基数（ 10） 后查看余数。例如，153 % 10 是 3。要获取剩余的数字，你可以对 15 重复该过程， 15 是 153 // 10。这个过程适用于任何基数。唯一的问题是要以相反的顺序得到数字（从右到左）。当 num 小于 base 时会发生递归的基本情况，输出就是 num。在一般情况下，函数（递归）打印 num // base 的数字，然后打印 num % base。你应该在连续输出之间放置一个空格，因为基数大于 10 时，会打印出多个字符的“数字”。例如， baseConversion(1234, 16)应打印 4 13 2。